### DISCUSSION TITLE
Making Sure I Understand the Outputs of plot_predictions Properly

### DISCUSSION THREAD

**Discussion by tanyalemieux22** [vincentarelbundock/marginaleffects#1767] [Q&A]
Hello,
I did an acoustic telemetry study on 4 species of fish in 3 locations. The response variable I'm using for one of my analyses is seasonal residency which is total days detected in a location per season divided by total days detected in any location per season. This means my response variable needs a ordered beta model because it has no zeros but can have ones.

The model I've chosen for my analysis is:
`sr_mod5 <-
  glmmTMB(
    seas_res ~
      season + 
      location + 
      species +
      release +
      location * season * release +
      location * season * species +
      (1 | transmitter_id),
    data = seasonal_res,
    family = ordbeta(link = "logit"))`

seas_res is the residency values
location is the location the fish was detected in
release is where the fish was release after I tagged it
transmitter_id is the id of the acoustic tag I implanted in the fish

From this model, both of the 3-way interaction terms are significant so I'm plotting each using the code below. 
```
plot_predictions(
  sr_mod5,
  by = c("location", "season", "release"),
  vcov = TRUE)
```

```
plot_predictions(
  sr_mod5,
  by = c("location", "season", "species"),
  vcov = TRUE)
```

<img width="2560" height="1440" alt="Screenshot 2026-09-13 at 10 28 07 AM" src="https://github.com/user-attachments/assets/baf5d11e-661c-405e-b78c-403b160abd5d" />
<img width="2560" height="1440" alt="Screenshot 2026-09-13 at 10 27 52 AM" src="https://github.com/user-attachments/assets/3796560c-ba39-4e42-bd77-29764153e026" />

I'll make the graphs pretty later but what I want to make sure I understand properly is the estimate. Would the estimate be the predicted residency (e.g., the predicted residency for white sucker in the river in fall is 0.3)? Or, would the estimate be the predicted probability that residency is 1 [e.g., there's a 30% chance that white sucker is a full resident of the river (i.e., residency = 1) in the fall]? 

Thanks!!

Tanya

**Comment by vincentarelbundock** [comment]
Hi, thanks for posting!

Unfortunately, I'm not an expert here, so I don't think I can help. But I'll tag @saudiwin since he'll likely know the answer.

Also, loading the `/marginaleffects` skill in Claude gives us an interesting explanation which, at first glance, seems very reasonable. Especially note (1) the key is to figure out what the scale of the default `type="response"` in the `predict.glmmTMB()` method supplied by that package, and (2)  the uncertainty issue mentioned at the end is quite important.

>   Short answer: neither. The estimate is on the residency scale (0–1), but it is the mean of the continuous part of the model
>   only — roughly "predicted residency among fish that are neither absent nor fully resident." It is not the full expected
>   residency, and it is not P(residency = 1).
> 
>   Why. An ordered beta model has three pieces driven by one linear predictor η and two cutpoints ψ₁ < ψ₂:
> 
>   - P(Y = 0) = 1 − plogis(η − ψ₁)
>   - P(Y = 1) = plogis(η − ψ₂)
>   - P(0 < Y < 1) = plogis(η − ψ₁) − plogis(η − ψ₂), with E[Y | 0 < Y < 1] = plogis(η)
> 
>   The quantity you almost certainly want is the unconditional mean:
> 
>   E[Y] = P(Y=1)·1 + P(0<Y<1)·plogis(η)
> 
>   But predict.glmmTMB(type = "response") for ordbeta returns plogis(η) — it ignores the cutpoints entirely ("response" and
>   "conditional" return identical values for this family). marginaleffects passes type = "response" straight through to
>   predict(), so plot_predictions() inherits that behavior. This is a glmmTMB limitation, not a marginaleffects choice.
> 
>   Confirmation on simulated data with known truth (glmmTMB 1.1.15, n = 2000, true cutpoints −2.5 and 1.0, two groups):
> 

>   ┌──────────────────────────────────┬─────────┬─────────┐
>   │             quantity             │ group A │ group B │
>   ├──────────────────────────────────┼─────────┼─────────┤
>   │ avg_predictions(type="response") │ 0.579   │ 0.713   │
>   ├──────────────────────────────────┼─────────┼─────────┤
>   │ E[Y], full expected residency    │ 0.657   │ 0.789   │
>   ├──────────────────────────────────┼─────────┼─────────┤
>   │ observed mean(y)                 │ 0.651   │ 0.789   │
>   ├──────────────────────────────────┼─────────┼─────────┤
>   │ P(Y = 1)                         │ 0.395   │ 0.546   │
>   ├──────────────────────────────────┼─────────┼─────────┤
>   │ observed proportion y == 1       │ 0.413   │ 0.533   │
>   └

──────────────────────────────────┴─────────┴─────────┘
> 
>   So type = "response" understates expected residency, and the gap grows with the share of 1s — which is exactly your
>   situation, since your outcome has no zeros but plenty of ones.
> 
>   What I'd recommend. Refit with the ordbetareg package, which fits the same model through brms:
> 
>   library(ordbetareg)
> 
>   sr_mod5 <- ordbetareg(
>     seas_res ~ season + location + species + release +
>       location * season * release + location * season * species + (1 | transmitter_id),
>     data = seasonal_res)
> 
>   plot_predictions(sr_mod5, by = c("location", "season", "release"))
>   plot_predictions(sr_mod5, by = c("location", "season", "species"))
> 
>   ordbetareg's posterior_epred() returns the full E[Y] including the point masses, so plot_predictions() and
>   avg_predictions() give you the correct expected-residency scale directly, with proper posterior intervals rather than a
>   delta-method approximation. It also handles the random effects coherently: with brms, marginaleffects propagates
>   uncertainty from the whole posterior instead of fixed effects only.
> 
>   If you want the other quantity as well — "probability that a fish is a full resident," P(Y = 1) — that's available from the
>   brms fit too, and with three-way interactions those two quantities can tell genuinely different stories. Happy to sketch
>   that if it's useful.
> 
>   One smaller note on your current call: you didn't set re.form, so glmmTMB defaults to re.form = NULL and the predictions
>   include each transmitter_id's estimated random effect, while the standard errors reflect only fixed-effect uncertainty
>   (that's what the marginaleffects warning is about). For population-level statements add re.form = NA; if you keep re.form =
>   NULL, treat the intervals as too narrow. Also, vcov = TRUE is already the default, so you can drop it.

**Comment by tanyalemieux22** [comment]
Hello! Thanks for your response!

I looked into the ordbetareg package and my preference would be not to go that route since it uses Bayesian methods. 

Do you know how to have glmmTMB and plot_predictions consider the random effects properly? Or is that what is meant by re.form = NA? I need the model to consider the random effect of fish id in case that's impacting residency. But ultimately, I do want to make population level inferences I think.

I found this post https://github.com/vincentarelbundock/marginaleffects/issues/1064, but honestly it's way over my head and I can't figure out how to get it to work properly. 

Re your comment "(1) the key is to figure out what the scale of the default type="response" in the predict.glmmTMB() method supplied by that package", I'm sorry but I'm not sure what you mean by that. Would you be able to state it another way please?

I'm also confused about what the above blurb is saying about what my estimate means because I didn't use predict.glmmTMB at all. Is it saying that because the marginaleffects does that in the background?

Thanks!

**Reply by vincentarelbundock** [reply]
What I meant is that by default, all `marginaleffects` functions pass a `type="response"` to the `predict()` method supplied by the model-fitting package (here: `glmmTMB`). That `type` argument determines the scale of the predictions, which obviously matters a lot for interpretation (think "link" vs. "response" scales in a simple logit model; obviously it matters whether we make predictions on the probabiity or log-odds scale).

The other issue raised above (related to `re.form`) was about the stndard errors, not the main predictions. The predictions should be fine, but the standard errors may not be the ones you are after, since they ignore the uncertainty related to the random components of your model.

AFAIK, there no "canned" solution in R (or any other language) to take that kind of uncertainty into account. It is a complicated problem, and the only "solution" I'm aware of is to go bayesian.

**Reply by vincentarelbundock** [reply]
On `type`, what I recommend is to look at the `glmmTMB` documentation to figure out what `predit(model, type="response")` does. Then predictions from `marginaleffects` should be expressed on the same scale.

**Comment by saudiwin** [comment]
The estimate produced by marginal effects for ordered beta (glmmTMB or ordbetareg) is the average proportion for a given value of covariates (or averaged over the whole dataset if no covariates are specified).

Specifically, it is Pr(Y=0) * 0 + Pr(0-1) * Beta(Y) + Pr(1) * 1. Given that you have no zeroes, it will boil down to the ratio between continuous responses and the number of 1s.

the issues with incorporating uncertainty from random effects in MLE fits is well-known; however, for predictive purposes (such as predicting proportions in a new area or species) it should work reasonably well. The glmmTMB package is set up to permit the best possible inference with random effects in an MLE framework. 

@bbolker would have more info & also of course the glmmTMB docs

I don't think there is an "issue" here @vincentarelbundock if you feel like closing this one.

**Reply by saudiwin** [reply]
Ah nm this is a "discussion", I need to keep up with the times 😅

**Reply by vincentarelbundock** [reply]
Thanks Bob, I really appreciate this!

**Comment by tanyalemieux22** [comment]
Hello,
Thanks @vincentarelbundock and @saudiwin for your time and your responses. I'm sorry but I'm still confused about what you're saying.

> What I meant is that by default, all marginaleffects functions pass a type="response" to the predict() method supplied by the model-fitting package (here: glmmTMB). That type argument determines the scale of the predictions, which obviously matters a lot for interpretation (think "link" vs. "response" scales in a simple logit model; obviously it matters whether we make predictions on the probabiity or log-odds scale).

For this comment, I found this: https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/man/predict.glmmTMB.Rd (which honestly is again over my head) but looking at it, it seems like I'd have to tell it which `type` to choose. So would that mean that when the marginaleffects package is passing a `type` to it as you said, then it's sending `response` as the `type` if that's what I entered in plot_predictions or avg_predictions and therefore the estimate would then be **_average predicted residency_** since my response variable is residency and the plot_predictions function does the same thing as the avg_predictions function in that it averages what you enter when the code is in that format? Also because @saudiwin did mention glmmTMB does include 0s, continuous, and 1s if present.

> The other issue raised above (related to re.form) was about the stndard errors, not the main predictions. The predictions should be fine, but the standard errors may not be the ones you are after, since they ignore the uncertainty related to the random components of your model. AFAIK, there no "canned" solution in R (or any other language) to take that kind of uncertainty into account. It is a complicated problem, and the only "solution" I'm aware of is to go bayesian.

For this comment, I think I may have an issue because I want to compare the levels in the fixed effects (e.g., whether location * season * release is significantly different from each other combination) and my understanding is that if the standard errors are wrong then the p-values could also be wrong. The code I used for this is:
`avg_predictions(
  sr_mod6,
  type = "response",
  by = c("location", "season", "release"),
  hypothesis = difference ~ pairwise)`

`avg_predictions(
  sr_mod6,
  type = "response",
  by = c("location", "season", "species"),
  hypothesis = difference ~ pairwise)`

Though the second chunk kept giving me the error:
`Error: This command will generate many estimates. Set options(marginaleffects_safe=FALSE) to circumvent this guardrail.`
I didn't turn anything off there yet though because I wanted to make sure I understood what I was doing. As a side question, is it ok to set that option to silence the error?

Also, I don't see`re.form = NA` as an option for any function I'm using ( https://github.com/glmmTMB/glmmTMB/blob/master/glmmTMB/man/glmmTMB.Rd, or https://github.com/vincentarelbundock/marginaleffects/blob/main/r/man/plot_predictions.Rd - I can't find the code for avg_predictions). So, if that piece of code does need to be used somewhere I'm not sure where to put it.

So my understanding of everything that has been said is that the p-values I get doing the pairwise comparisons could be wrong then and so will the error bars on my plots and there's no way to 100% get around that other than going Baysian. However, based on what you both said the predictions will be correct and for what I'm modeling the standard errors will likely be good enough. Is that right?

Also, I really want to figure out how to do this with random effects because I have a lot of modelling to do and I'm sure what I mention below will not apply in many of the other models I'm doing.

But I did find in this particular model the impact of my random effect is pretty much zero, so I'm thinking that helps.
`summary(sr_mod6)
 Family: ordbeta  ( logit )
Formula:          seas_res ~ season + location + species + release + location *  
    season * release + location * season * species + (1 | transmitter_id)
Data: seasonal_res
      AIC       BIC    logLik -2*log(L)  df.resid 
   -223.4      50.3     175.7    -351.4       468 
Random effects:
Conditional model:
 Groups         Name        Variance  Std.Dev. 
 transmitter_id (Intercept) 6.208e-12 2.492e-06
Number of obs: 532, groups:  transmitter_id, 55`

So since this is the case, for this particular model I would think that means it's more likely the standard errors are indeed correct?

@saudiwin Sorry, after writing all that, I feel like you maybe said in your comment that all is well and to just go ahead with what I did. I just really want to understand everything and what I'm doing to make sure what I'm publishing is accurate and so I know what I'm talking about in my PhD defence.

Thanks for your patience!

Tanya

**Comment by bbolker** [comment]
A few things here (coming in late):

* "I just really want to understand everything and what I'm doing to make sure what I'm publishing is accurate and so I know what I'm talking about in my PhD defence." is entirely understandable!
* it does look like you have a singular fit (random-effect variance $\approx 0$), so you can drop that term from your model and not worry about the mixed-model specific stuff (dropping a negligible term will have no effect on the predicted values, although it may slightly underestimate the uncertainty in some cases &mdash; Bayesian approaches will do a better job here, although I can appreciate why you might not want to go down that rabbit hole)

The rest is a little more complicated. Bottom line, I think @saudiwin is wrong (or I'm misunderstanding what he's saying) and Claude is right: `predict(., type = "response")` and `predictions(.)` **are giving the mean value for non-zero, non-one values**, i.e. not taking the 0/1 censoring into account.

```r
library(glmmTMB)
library(marginaleffects)

set.seed(101)
dd <- as.data.frame(matrix(ncol = 0, nrow = 1000))
dd$y <- simulate_new(~1, family = ordbeta,
                     newdata = dd,
                     newparams = list(beta = 0.5, betadisp = -2,
                                      psi = c(-1,-1)))[[1]]
m <- glmmTMB(y ~ 1, data = dd, family = ordbeta)
p0 <- predict(m, type = "response")[[1]]

par0 <- m$obj$env$parList()

ae <- function(x,y) all.equal(x,y, tolerance = 2e-3)

## prob(0)
ae(mean(dd$y==0), plogis(par0$psi[1]-par0$beta))
## prob(1)
ae(mean(dd$y==1), plogis(par0$beta-par0$psi[2]))
## mean(y[0<y<1])
ae(plogis(par0$beta), mean(subset(dd, y>0 & y<1)$y))
p1 <- suppressWarnings(predictions(m)$estimate[[1]])
p2 <- predict(m, type = "response")[[1]]
ae(p1, p2)
ae(p1, plogis(par0$beta))
```

I think that `glmmTMB` probably needs to do better in providing options for ordbetareg-prediction. `?predict.glmmTMB` outlines some of the choices that are relevant to zero-inflated models (the closest analogue to ordbetareg models:

> "response"
> expected value; this is mu*(1-p) for zero-inflated models
> and ‘mu’ otherwise

> "conditional"
> mean of the conditional response; ‘mu’ for all models
> (i.e., synonymous with ‘"response"’ in the absence of
> zero-inflation

> "zprob"
> the probability of a structural zero (returns 0 for
> non-zero-inflated models)

> "zlink"
> predicted zero-inflation probability on the scale of the
> logit link function (returns ‘-Inf’ for non-zero-inflated
> models)

This suggests that we should change `predict(., type = "response")` to return the mean across all values (i.e. `0*prob(0) + x*prob(0<x<1) + 1*prob(1)`), in parallel with what's done for zero-inflated models, and make `type="conditional"` return what's *currently* returned by `type="response"`.  With a bit of an abuse of notation, "zprob" could return a two-column matrix (`cbind(prob(0), prob(1))`), and "zlink" could return a two-column matrix with the logits ... (this should probably be opened as a `glmMTMB` issue ...)

**Comment by saudiwin** [comment]
Yes I think the default method should be the conditional expectation for the outcome rather than a sub-component. Of course not my package, though.

**Comment by tanyalemieux22** [comment]
Sorry for the delayed response. @bbolker is that code you provided what glmmTMB is doing in the background? 

Re getting glmmTMB updated to make predictions based on including the 0s, continuous data, and 1s, would I need to open an issue to get that done? Also, do you know how long roughly it takes to make updates to packages when people request them?

Also, just want to make sure I 1000% understand this part....When I run glmmTMB() with family = ordbeta(link = "logit")) the p-values etc. I get from the model will be correct...it's just that when I then use any of the marginaleffects package functions on that model, the standard errors I get for the marginal effects will be wrong if my model has random effects, and since the standard errors could be wrong the p-values of the marginal effects could be wrong too?

**Comment by tanyalemieux22** [comment]
Apologies for the second post...I'm working on another of my analyses and I have the same initial question I had in this post about this analysis.

For this one my response is binary presence/absence and my model is below. In this analysis I'm specifically looking at what impacts northern pike usage of one location.

`np_bscp_mod4 <-
  glmmTMB(
    presence_binary ~
      scale(total_length_mm) +
      release +
      scale(muskellunge) +
      scale(northern_pike) +
      scale(walleye) +
      scale(white_sucker) +
      scale(DO_percent_sat) +
      scale(Temperature) +
      scale(avg_discharge_m3s) +
      scale(do_diff) +
      scale(discharge_diff) +
      scale(avg_discharge_m3s) * scale(DO_percent_sat) +
      scale(avg_discharge_m3s) * scale(Temperature) +
      (1 | transmitter_id),
    data = np_bscp,
    family = binomial(link = "logit"))`

As an example, I'm plotting the impact of temperature with the below code and get the below image.

`plot_predictions(
    np_bscp_mod4,
    condition = "Temperature",
    type = "response",
    vcov = TRUE,
    conf_level = 0.95)`

<img width="1197" height="622" alt="image" src="https://github.com/user-attachments/assets/7baae422-fc99-42a5-98fb-7d65e2905153" />

In this situation, would the estimate be the probability that northern pike would be present e.g., at 5C there's a 10% chance they'd be present?
